### PR TITLE
Detect to ServiceMonitor and PrometheusRule CRD

### LINKED ISSUE #6551: Reconcile and enable monitoring even after virt-operator startup
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind enhancement

**What happened**:

virt-operator deteremine whether or not to enable monitoring by checking the existence of sevicemonitor CRD. However it only occurs [during startup]. This is somehow limited since a user may deploy prometheus monitoring stack after virt-operaetor startup.

[during startup]: https://github.com/kubevirt/kubevirt/blob/5a22acb1868e17824df34044a60a5e42b91b9357/pkg/virt-operator/application.go#L239-L268 

**What you expected to happen**:

Whenever a user deploy prometheus monitoring CRD, virt-operator must enable monitoring functionality.

**How to reproduce it (as minimally and precisely as possible)**:

1. Deploy KubeVirt in a cluster
2. Wait a moment
3. Deploy prometheus-operator CRD. Then there is no servicemonitor for KubeVirt being created.

**Environment**:
- KubeVirt version (use `virtctl version`): v0.45.0

### PR DESCRIPTION
**Pull request by chrisho**
**What this PR does / why we need it**:
Reinitialize when detected to ServiceMonitor and PrometheusRule CRD has been introduced/removed.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6551

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by guangbochen** [comment]
cc @vasiliy-ul, can u please help to take a review of the PR, thanks.

**Review by vasiliy-ul — COMMENTED** [review]
Hi @chrisho, @guangbochen. I had a quick look and the code seems okay. I think it would be good to have it covered by some tests. Perhaps testing e2e will be a bit tricky. Maybe we could add a unit test then? Smth similar to the one below, wdyt?

https://github.com/kubevirt/kubevirt/blob/2baf5303dddfe71c98a2510357c464a8f7bdd3ca/pkg/virt-controller/watch/application_test.go#L246

@xpivarc, @davidvossel, could you also take a look here? If I understand it right, it implements the suggested approach https://github.com/kubevirt/kubevirt/issues/6551#issuecomment-1195972383 that you guys have agreed upon.

/ok-to-test

**Comment by chrisho** [comment]
> Hi @chrisho, @guangbochen. I had a quick look and the code seems okay. I think it would be good to have it covered by some tests. Perhaps testing e2e will be a bit tricky. Maybe we could add a unit test then? Smth similar to the one below, wdyt?
> 
> https://github.com/kubevirt/kubevirt/blob/2baf5303dddfe71c98a2510357c464a8f7bdd3ca/pkg/virt-controller/watch/application_test.go#L246
> 
> @xpivarc, @davidvossel, could you also take a look here? If I understand it right, it implements the suggested approach [#6551 (comment)](https://github.com/kubevirt/kubevirt/issues/6551#issuecomment-1195972383) that you guys have agreed upon.
> 
> /ok-to-test

Thanks, @vasiliy-ul.
The unit tests will be added later.

File: pkg/virt-config/configuration.go [hunk 1]
```diff
@@ -371,7 +380,37 @@ func (c *ClusterConfig) HasDataVolumeAPI() bool {
 	objects := c.crdInformer.GetStore().List()
 	for _, obj := range objects {
 		if crd, ok := obj.(*extv1.CustomResourceDefinition); ok && crd.DeletionTimestamp == nil {
-			if isDataVolumeCrd(crd) {
+			if isServiceMonitor(crd) {
```
Comment (vasiliy-ul) [inline pkg/virt-config/configuration.go:383]: This change does not look right for me. `HasDataVolumeAPI` should check for DataVolume CRD with `isDataVolumeCrd`

File: pkg/virt-operator/application_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	. "github.com/onsi/gomega"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/testutils"
+)
+
+var _ = Describe("Application", func() {
+	Describe("Reinitialization conditions", func() {
```
Comment (vasiliy-ul) [inline pkg/virt-operator/application_test.go:35]: Probably this `Describe` block is redundant here

File: pkg/virt-operator/application_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			if addPrometheusRuleCrd {
+				testutils.AddPrometheusRuleAPI(crdInformer)
+			} else if removePrometheusRuleCrd {
+				testutils.RemovePrometheusRuleAPI(crdInformer)
+			}
+
+			objects := crdInformer.GetStore().List()
+			for _, obj := range objects {
+				if crd, ok := obj.(*extv1.CustomResourceDefinition); ok && crd.DeletionTimestamp == nil {
+					GinkgoWriter.Println(crd.Spec.Names.Kind)
```
Comment (vasiliy-ul) [inline pkg/virt-operator/application_test.go:66]: This looks like a debug print. Perhaps its not needed for the final version.

File: pkg/virt-operator/application_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+				}
+			}
+
+			app.clusterConfig.SetConfigModifiedCallback(app.configModificationCallback)
+
+			GinkgoWriter.Printf("%t %t\n", hasPrometheusRules, app.clusterConfig.HasServiceMonitorAPI())
```
Comment (vasiliy-ul) [inline pkg/virt-operator/application_test.go:72]: Same here: probably the debug print is not needed.

**Comment by vasiliy-ul** [comment]
@chrisho, it also looks like you have some formatting issues. You need to run `make && make test` and then commit the changes added by the formatter.

**Review by vasiliy-ul — APPROVED** [review]
Looks good now. Thanks!

**Comment by guangbochen** [comment]
@dhiller @xpivarc @davidvossel  can u PTAL of the PR, thanks.

**Comment by vasiliy-ul** [comment]
For the record: https://kubernetes.slack.com/archives/C0163DT0R8X/p1683722555609669

The implementation here has been discussed in the linked issue, and it follows the suggested approach https://github.com/kubevirt/kubevirt/issues/6551#issuecomment-1195972383. From my perspective, the PR is okay to get in.

/approve

**Pull request merged**